### ISSUE TITLE
RequireThis: False positive for lambda parameters

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#4207]
````
$ cat TestClass.java
class InputAllowLocalVars {

    private String s1 = "foo1";

    void foo1() {
        final java.util.List<String> strings = new java.util.ArrayList<>();
            strings.add("foo1");
            strings.stream().filter(s1 -> s1.contains("f"))
            .collect(java.util.stream.Collectors.toList());
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="RequireThis">
            <property name="checkMethods" value="false" />
            <property name="validateOnlyOverlapping" value="false" />
        </module>
    </module>
</module>


$ java -jar checkstyle-7.6.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:8:37: Reference to instance variable 's1' needs "this.". [RequireThis]
[ERROR] TestClass.java:8:43: Reference to instance variable 's1' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 2 errors.
````

http://rveach.no-ip.org/checkstyle/checkstyle.php?action=view&config=2b75d3a2ecffeba47690e0416ccd9906&code=c5b8991fb27367d9bfbf7d120993a7cb&checkstyle=checkstyle-7.6.1-all.jar&printTree=

----------------------------
Expected: no violations as ```this``` cannot be used for ```s1``` on line 8 and column 37 (compile time error); ```s1``` on line 8 and column 43 is a lambda parameter and it does not make sense to prefix it with ```this```.

Note, that if ```validateOnlyOverlapping``` is set to true, the violations disappear.

**Comment by ps-sp** [comment]
I am on it.

**Comment by ps-sp** [comment]
// The following statement causes the compiler to generate
            // the error "local variables referenced from a lambda expression
            // must be final or effectively final" in statement A:
            //
            // x = 99;
            
            Consumer<Integer> myConsumer = (y) -> 
            {
                System.out.println("x = " + x); // Statement A
                System.out.println("y = " + y);
                System.out.println("this.x = " + this.x);
                System.out.println("LambdaScopeTest.this.x = " +
                    LambdaScopeTest.this.x);
            };

We wouldn't need `RequireThis` to report another error for statement A, right.

**Comment by ps-sp** [comment]
class foo {   
   
         int x;

         void methodInFirstLevel(int x) {
            
             //There would be no compiler errors for the statement A
             //because the parameter x is effectively final.
             
             Consumer<Integer> myConsumer = (y) -> 
             {
                 System.out.println("x = " + x);         // Statement A
                 System.out.println("y = " + y);
                 System.out.println("this.x = " + this.x);
                 System.out.println("LambdaScopeTest.this.x = " +
                     LambdaScopeTest.this.x);
             };

             myConsumer.accept(x);

         }
    }

Would it be desirable for `RequireThis` to report an error for statement A ?

**Comment by ps-sp** [comment]
Also, **lambda expressions don't introduce a new level of scoping and hence don't have any shadowing issues**. All they can access is `final` or _effectively_ `final` local variables and members of the enclosing classes. Hence the user/programmer would/should take that into account and code in the exact way he wants to refer to the variables. Hence, `RequireThis` violations thrown for variables inside lambdas might be unnecessary ! 

So should any violations be thrown at all from inside of lambdas by `RequireThis` ? I can't think of any situation or weird code for which it will be helpful to have violations from `RequireThis` from inside lambdas.

@MEZk @romani @rnveach Please address. Thank you.

**Comment by rnveach** [comment]
> Would it be desirable for RequireThis to report an error for statement A ?

Since it is a class field, yes, without a doubt.
This check probably shouldn't concern itself with if the variable is `final` (unless someone can provide a reason). Just is it a class field/method being referenced.

> Hence, RequireThis violations thrown for variables inside lambdas might be unnecessary !
>  I can't think of any situation or weird code for which it will be helpful to have violations from RequireThis from inside lambdas.

Purpose of check is to place `this` on class fields and methods. Yes, code will compile without it but it is about clearly showing the variable being referenced is not a local variable. Like defining variables as `final`, it allows you to see more of the purpose of the variable without scanning nearby/whole code.

If variable/method being referenced in the lambda can have `this` added to it, it should be reported (unless `validateOverlapping` gives it a reason to not report something).

**Comment by MEZk** [comment]
@ps-sp 
> there's no violation for statement A because it's previous sibling isn't null and neither it's parent an assignment token.

Yes, this is not an assignment.

> Also, in the current form, RequireThis points a violation for statement B. That is violation is acceptable, right ?

Yes, this is an assignment.

**Comment by rnveach** [comment]
Fix is merged
